### ISSUE TITLE
Issue using term files with filter command

### ISSUE THREAD

**Issue by zoependlington** [ontodev/robot#401]
Hello!

I have been using the filter command to create a slim as suggested in another ticket (#263). I received a request to develop a way to extract a desired subset minus any terms deemed "unwanted", i.e. a simplification of a subset. Both the subset and the unwanted terms are unfixed, relying heavily on text files as inputs and term files.

The filter command I'm using is modified from the ticket above as I have a list of unwanted terms to remove from the ontology, but wanting to preserve the ancestors, descendants and the hierarchy. Here I have a MIREOT extract subset of CHEBI as my input (using nicotinic antagonist as my extract term), have a text file with unwanted terms (unwanted.txt) and the output remove.owl:

`robot filter --input chebi-annotated.owl --term-file unwanted.txt --select "ancestors descendants annotations" --output remove.owl`

I'm testing @rctauber's "gaps" branch as I want to preserve the structure. It works wonderfully when I test it with one term in my term file (neurotransmitter agent). It removes the term and preserves the structure in exactly the way I'd like. My original file is on the left, the result of the filter command (remove.owl) is on the right:

![image](https://user-images.githubusercontent.com/36990116/48205446-26ae6c00-e364-11e8-8c12-c05ae25ceaef.png)

However, when I add another term (pharmaceutical) to the unwated.txt file so it now contains the two IDs on separate lines:
CHEBI:52217
CHEBI:35942

The command ceases to remove anything (again, original on the left, the remove.owl on the right):

![image](https://user-images.githubusercontent.com/36990116/48205556-55c4dd80-e364-11e8-8049-ba3495616228.png)

To check this was not due to the version I was using, I tested this using ROBOT version 1.2.0-alpha-1. Using the --term-file command, even if the unwanted.txt contained only one term, resulted in no removal of the desired neurotransmitter agent term i.e. no action:
![image](https://user-images.githubusercontent.com/36990116/48205957-24004680-e365-11e8-8e2c-dfdae3e47e8b.png)

While `robot filter --input chebi-annotated.owl --term CHEBI:35942  --select "ancestors descendants annotations" --output remove.owl` expectantly produced the removal of neurotransmitter agent (even without preservation of structure):

![image](https://user-images.githubusercontent.com/36990116/48206006-47c38c80-e365-11e8-858e-e24da6c37b2b.png)

In other instances testing the filter command, I have also been unable to use the --term-file option, but have been able to use multiple --term options in the same filter command. However, this seems to not be working in this situation (which I'm guessing is due to not selecting "self"? This is the only difference between my two pipelines using filter) with either version of ROBOT.

Am I doing something wrong for the term file not to work? Any other suggestions for this problem?

Thank you!

**Comment by rctauber** [comment]
Hi @zoependlington - could you include the `unwanted.txt` term file so I can test it out?

If you're trying to remove the "unwanted" terms plus their ancestors and descendants, the `remove` command would probably be best for that (and include `self` in the select statement). Sorry if I'm not understanding this correctly.

**Comment by rctauber** [comment]
Whoops, didn't mean to close, sorry!

**Comment by zoependlington** [comment]
Here is the file with both test terms:
[unwanted.txt](https://github.com/ontodev/robot/files/2562379/unwanted.txt)

I was testing out the remove command a little while ago and believe that I ran into a similar issue, but I can't be sure. The problem is only wanting to remove the unwanted term while keeping the ancestors and descendants and filling the gap left by the removed term. This filter command with the preservation I was testing today worked wonderfully for that when removing one term. I may need to revisit trying out remove! 🙂Just wanted to let you know about the term file problems I'd noticed today.

Thank you Rebecca!

**Comment by rctauber** [comment]
Ah, I see! So you want to remove *only* the terms in the unwanted file and keep their ancestors and descendants. If you use the most up-to-date version of #263, the following should work:
```
robot remove --input chebi.owl --term-file unwanted.txt --output chebi-removed.owl
```

**Comment by zoependlington** [comment]
Yes, exactly! Fortunately, this remove command does remove the unwanted terms, but, unfortunately, does not preserve the structure even with `--preserve-structure true` added (although if I'm remembering rightly this is the default?), whereas when removing the terms individually via the filter method it was preserving the structure.

![image](https://user-images.githubusercontent.com/36990116/48210908-3cc22980-e370-11e8-87da-9caa79d473a4.png)

**Comment by rctauber** [comment]
I found the issue with `filter` you were having before... You're getting the `ancestors` and the `descendants` of those two terms. "pharmaceutical" (`CHEBI:52217`) is a parent of "neurotransmitter agent" (`CHEBI:35942`), so they're getting pulled back in by each other (ancestors of "neurotransmitter agent" includes "pharmaceutical" and descendants of "pharmaceutical" include "neurotransmitter agent"). I hope I explained that okay 😅 

I'll look more into why the "cholinergic drug" is being orphaned.

**Comment by zoependlington** [comment]
Ah, yes! That makes a lot of sense now. I was completely blind to that possibility - oops! Thank you! 

I did run two separate filter commands (just to check) which gave the output I was expecting, although would not work for the unpredictable scope of this request. The result that I'm expecting is on the right compared to the original on the left:

![image](https://user-images.githubusercontent.com/36990116/48213559-07b8d580-e376-11e8-8916-de3bdf2f53a9.png)

Looking at the bigger picture for this request from my user now, using remove makes a lot more sense as the extract from CHEBI could be a lot more complex, with a large number of unwanted terms. Removing these individual terms while preserving the hierarchy would be less prone to error than filtering based on ancestors and descendants of unwanted terms. 

Thanks for your help so far! 😁

**Comment by rctauber** [comment]
Got it! I also found the cause of the incorrect gap-filling for `remove`. The gap-filling method looks at the terms that were included in the set and creates the relationships between those. In `filter`, this works perfectly, because your set is made up of the terms you want to keep. In `remove`, though, your set is the *opposite* of the terms you want to keep. So for the gap-filling to work with `remove`, we needed to get the opposite of the set of terms that we removed and use that to fill the gaps.

I just pushed these changes to #357. Thank you SO much for finding this! If you are able to build from the latest commit on that branch and see if it works for you, I would really appreciate it.
